### ISSUE TITLE
error when running phyluce_assembly_match_contigs_to_probes

### ISSUE THREAD

**Issue by SusanaFreitas** [faircloth-lab/phyluce#48]
Hello!
I have been trying to run phyluce_assembly_match_contigs_to_probes and after testing some different probe name formats I still can't make it to run...
I got this error:
Traceback (most recent call last):
  File "/usr/local/packages6/apps/binapps/conda/phyluce/bin/phyluce_assembly_match_contigs_to_probes", line 345, in <module>
    main()
  File "/usr/local/packages6/apps/binapps/conda/phyluce/bin/phyluce_assembly_match_contigs_to_probes", line 248, in main
    uces = set(new_get_probe_name(seq.id, regex) for seq in SeqIO.parse(open(args.probes, 'rU'), 'fasta'))
  File "/usr/local/packages6/apps/binapps/conda/phyluce/bin/phyluce_assembly_match_contigs_to_probes", line 248, in <genexpr>
    uces = set(new_get_probe_name(seq.id, regex) for seq in SeqIO.parse(open(args.probes, 'rU'), 'fasta'))
  File "/usr/local/packages6/apps/binapps/conda/phyluce/bin/phyluce_assembly_match_contigs_to_probes", line 237, in new_get_probe_name
    return match.groups()[0]
AttributeError: 'NoneType' object has no attribute 'groups'

And my probes file looks like:

> uce-500_1
> TCGAGTACAGTGGTTACTTCAGAATTTGAGGCTTAGTCGTCTTCTTAAAGGTTAGAAACTCTGTT
> uce-500_2
> TGTCTGACACTTGAGAAGACTATTTTGGCATATCACATTGTCGAGTACAGTGGTTACTTCAGAAT
> uce-500_3
> GCTGTAGCATAAGATGCCTTTTTATTGTGCTGCATCAGCCGCTAAGTGATGAGGATGCTATCAG

Am I doing something wrong? Is the problem due to the probe files or something else?

Thanks in advance!!!
Susana

**Comment by brantfaircloth** [comment]
Susana, do things work with the test data/and probe sets?  what you are trying to do is not the standard workflow - you'll need to create a reasonable regular expression to do what you want to do.  I'll try and take a look soon...

**Comment by SusanaFreitas** [comment]
aham! I didn't notice the "p", sorry about that.
I just tested again with both the corrected fasta file (see bottom) and the test probe set and if still fails. This time the error message is (for both):
Traceback (most recent call last):
  File "/usr/local/packages6/apps/binapps/conda/phyluce/bin/phyluce_assembly_match_contigs_to_probes", line 345, in <module>
    main()
  File "/usr/local/packages6/apps/binapps/conda/phyluce/bin/phyluce_assembly_match_contigs_to_probes", line 259, in main
    uces
  File "/usr/local/packages6/apps/binapps/conda/phyluce/bin/phyluce_assembly_match_contigs_to_probes", line 130, in create_probe_database
    raise sqlite3.OperationalError("Cannot create database")
sqlite3.OperationalError: Cannot create database

The assembly was performed using phyluce_assembly_assemblo_velvet, so I assume everything is correct in the folder structure (that I did not change).

Sorry again about the change of question.

Thanks!
Susana



probes fasta file
>uce-500_p1
TCGAGTACAGTGGTTACTTCAGAATTTGAGGCTTAGTCGTCTTCTTAAAGGTTAGAAAC
>uce-500_p2
TGTCTGACACTTGAGAAGACTATTTTGGCATATCACATTGTCGAGTACAGTGGTTACT
>uce-500_p3
GCTGTAGCATAAGATGCCTTTTTATTGTGCTGCATCAGCCGCTAAGTGATGAGGATGCTA
>uce-500_p4
TTCTTGTTAAAGCTATGATTTTGTCATCCTCTTGCTTGTATGAGCAAGGTAAGTGAATAATAG
>uce-500_p5
TGTGTCCTTTAGTACATTATAATTTCAAACTCATTCATTCTGTCTGACACTTGAGAAGACTATT

**Comment by SusanaFreitas** [comment]
Hello,
I am sorry to insist but I still have the same error :(
I tried:
1) run the assembly with filename format anas_platyrhynchos_KGH2267
2) re-run with the same format (correcting folder names)
3) re-run the assemblies with the filename format anas_platyrhynchos1

All of the above led to the same error when running the phyluce_assembly_match_contigs_to_probes. Error message:

_### Traceback (most recent call last):
  File "/usr/local/packages6/apps/binapps/conda/phyluce/bin/phyluce_assembly_match_contigs_to_probes", line 345, in <module>
    main()
  File "/usr/local/packages6/apps/binapps/conda/phyluce/bin/phyluce_assembly_match_contigs_to_probes", line 259, in main
    uces
  File "/usr/local/packages6/apps/binapps/conda/phyluce/bin/phyluce_assembly_match_contigs_to_probes", line 130, in create_probe_database
    raise sqlite3.OperationalError("Cannot create database")
sqlite3.OperationalError: Cannot create database_


Log error message:

_###2016-02-24 20:17:02,285 - phyluce_assembly_match_contigs_to_probes - INFO - Creating the UCE-match database
2016-02-24 20:17:02,288 - phyluce_assembly_match_contigs_to_probes - CRITICAL - Database already exists
2016-02-24 20:17:02,288 - phyluce_assembly_match_contigs_to_probes - CRITICAL - Cannot create database
_

I also tried the example probes file and the error message was exactly the same.
Any help?

Thanks in advance.
Susana

**Comment by brantfaircloth** [comment]
Susana,  I'm afraid I'm not sure what your error actually is.  If you can run the tutorial data through as described in the tutorial, then that suggests the program is running as expected.  

From your earliest email, it appears you are using non-standard probe names and experimenting with input data initially processed using a different program.  I'm not entirely sure of this because i have no idea what the structure of the data is like that you are trying to process nor what the contigs are that you are trying to run the match_contigs_to_probes program against.

So, it's incredibly hard to help you track down your problem (which I'm not sure is a problem in the software)... Some idea of the directory structure you are working with, the command line call to match_contigs_to_probes you are using, etc, would go a long way towards making it easier to help you.

**Comment by SusanaFreitas** [comment]
Sorry, maybe I should have made that clearer...

> assemblies
I used velvet through phyluce (phyluce_assembly_assemblo_velvet) so the
folder structure should be as expected?

>probe names
My probe names are like this:
>uce-500_p1
ATGATGATGATGATG(...)
and I have been using this format for the last attempts.

> the assembly command line
phyluce_assembly_assemblo_velvet --config config.conf --output assemblies
--kmer 55 \
 --subfolder split-adapter-quality-trimmed --cores 25 --clean --log-path log

> the match contigs command line
phyluce_assembly_match_contigs_to_probes --contigs assemblies/contigs
--probes probes/pythonfile.fa --output ALIGN --log log

I don't really know what else to try (and I have invested so much time in
this that I wouldn't want to go back and make my own script now). Could it
be the velvet assembler?

Sorry, if there isnt anything more I can do (and that you can say about it)
I will try another way. I am just not to happy about quitting almost in the
end...

Many thanks!
Susana




2016-02-24 20:35 GMT+00:00 Brant Faircloth <notifications@github.com>:

**Comment by brantfaircloth** [comment]
Ok, so you have assemblies from velvet, and those appear to have worked.  can you email me several of the contig assembly files and the probe file you are using?

**Comment by JETitus** [comment]
Hi Susana,

It looks like the error you see below is because your symlinked contigs are not able to be read by sqilte3.

Log error message:

_###2016-02-24 20:17:02,285 - phyluce_assembly_match_contigs_to_probes - INFO - Creating the UCE-match database
2016-02-24 20:17:02,288 - phyluce_assembly_match_contigs_to_probes - CRITICAL - Database already exists
2016-02-24 20:17:02,288 - phyluce_assembly_match_contigs_to_probes - CRITICAL - Cannot create database

Check here: https://github.com/tobiashofmann88/Processing-Illumina-reads/wiki/UCE-tutorial

Basically your symlinked contigs, found in the directory called contigs nested within your assembly directory, will have the exact names as your sample directories. How squlte3 is going to need those symlinks in genus_species1.fasta. So if you have special characters like ?$#%^ or an extra underscore, the database will fail. You can simply change the name of the symlinks in the contigs folder and it should work. I hope that helps.

Cheers,
James

**Comment by brantfaircloth** [comment]
Actually, it appears that the error Susana was running across may have to do with Lustre file systems and SQLite (she was running her analyses on the cluster).  There is a somewhat well-known incompatibility between both, additional details of which you can track down from this list-serv posting: https://www.mail-archive.com/lustre-discuss@lists.lustre.org/msg03213.html.

**Comment by SusanaFreitas** [comment]
Hello!
I am sorry if what I am about to say will only increase the confusion...
Today I tested the run in another server with a different filesystem (NFS)
and it also didn't work. After James' email I tested the same script with a
subset of the assemblies and I did it without the symbolic links: the match
script ran with no errors! I was now about to run all assemblies without
symbolic links to see if that works, but my guess that maybe was the
problem.

However, I did not have any special characters that might explain the
problem. Also, because I tested it in a different server, I had to change
the paths of the soft links. So it was not also something on the path given
by the assembler.

I will do some tests (also with the symbolic links) and let you all know.

Many thanks!!
Susana

2016-02-25 19:03 GMT+00:00 Brant Faircloth <notifications@github.com>:

**Comment by brantfaircloth** [comment]
Okay, well belay what's above (although that can also cause a problem - so i'll leave it up there).  Properly configured NFS is not the issue - all of the machines I use run off of NFS (this spans about 5 totally different clusters as well as my own machines that I manage).

Are the symlinks that the program was initially trying to follow actually valid (meaning, do they point to the correct places)?  And, of course, when you move items to a different server, you'll always need to validate those (and any other) symlinks.

**Comment by SusanaFreitas** [comment]
They are valid (they lead to the correct file). I also tested (some of) the
symlinks I changed and they looked OK. Anyway, it is now running with all
the assembly files renamed and in the contigs folder (without any symlink).
I already have the probe.matches.sqlite and probe.matches.sqlite-journal,
as well as some of the *contigs.lastz files, so all appears to go
normally...

2016-02-25 19:25 GMT+00:00 Brant Faircloth <notifications@github.com>:

**Comment by davecarlon** [comment]
Hello and picking up on this thread. I'm getting the same error message trying to run phyluce_assembly_match_contigs_to_probes on my mac laptop (OS10.10.5) with the fish-uce-500 probes set. In this case I ran the 96 assemblies via the abyss script on a cluster, and then moved the resulting contigs/ folder over to my mac to try running phyluce locally. I assumed that the contigs/ folder contained the fasta files, but I see from this thread that the assembly script creates "soft" links to the contigs and fasta files. After I copy the /contigs folder onto my mac the links are lost and just the fasta files remain. SusanaFrietas' seems to have got this to work, but no luck here. Is there some other file that phyluce_assembly_match_contigs_to_probes is looking at to create the sqlite data table? Thx

**Error message:**

2016-11-14 15:12:40,523 - phyluce_assembly_match_contigs_to_probes - CRITICAL - Database already exists
2016-11-14 15:12:40,523 - phyluce_assembly_match_contigs_to_probes - CRITICAL - Cannot create database
Traceback (most recent call last):
  File "/Users/dcarlon/anaconda/bin/phyluce_assembly_match_contigs_to_probes", line 345, in <module>
    main()
  File "/Users/dcarlon/anaconda/bin/phyluce_assembly_match_contigs_to_probes", line 259, in main
    uces
  File "/Users/dcarlon/anaconda/bin/phyluce_assembly_match_contigs_to_probes", line 130, in create_probe_database
    raise sqlite3.OperationalError("Cannot create database")
sqlite3.OperationalError: Cannot create database

**Comment by brantfaircloth** [comment]
Sometimes this error raises it's head when the contig names have "weird" characters in them.  This could be something like starting with a number or containing punctuation (other than dash or underscore).

The symlinks to the files are not necessary, as long as the "contigs" folder contains files of the form:

```
genus-species1.contigs.fasta
genus-species2.contigs.fasta

```

Can you check to see if the contig names contain potentially problematic characters?

**Comment by davecarlon** [comment]
Like:
446_Sg_Baja:/mnt/research/dcarlon/clean-reads/446-Sg-Baja
447_Sg_Baja:/mnt/research/dcarlon/clean-reads/447-Sg-Baja
456_Sr_Baja:/mnt/research/dcarlon/clean-reads/456-Sr-Baja
....
All my contigs start with numbers. I'll switch label parts and see what happens!

**Comment by davecarlon** [comment]
That was it. I added a text prefix to all sample names.

**Comment by brantfaircloth** [comment]
Excellent - I need to add some error checking here to make this problem more clear...

**Comment by stuartwillis** [comment]
Hi Brant et al.,

We are experiencing the same error Susana described above:
Traceback (most recent call last):
  File "/home/lhunt/miniconda2/envs/phyluceENV/bin/phyluce_assembly_match_contigs_to_probes", line 345, in <module>
    main()
  File "/home/lhunt/miniconda2/envs/phyluceENV/bin/phyluce_assembly_match_contigs_to_probes", line 294, in main
    contig_name = get_contig_name(lz.name1)
  File "/home/lhunt/miniconda2/envs/phyluceENV/bin/phyluce_assembly_match_contigs_to_probes", line 232, in get_contig_name
    return match.groups()[0]
AttributeError: 'NoneType' object has no attribute 'groups'

The names are as they should be (genus_species1.contigs.fasta) and the symlinks work fine from the directory where the script is being executed ( Argument --contigs: /home/lhunt/miseq/uce/trinity_assemblies/contigs where XXX.contigs.fasta are found). We did quality trim the data and run Trinity from outside the phyluce pipeline so that we could run samples concurrently on our cluster. I'm wondering if there's some processing of the contig names in Trinity.fasta that is needed for lastz or sqlite. Interestingly, the script creates a .lastz file for the first sample with 4300+ entries and then craps out. I've suggested to Lizz to run the tutorial on our local workstation to make sure all the parts are working. Do you have any other thoughts?  Thanks! Stu

**Comment by brantfaircloth** [comment]
Hi Stuart,

This actually sounds like it may be a different problem.  Which assembly program did you use and what was the version number?  can you send me an example fasta header or two from the genus_species1.contigs.fasta file?  alternatively, you can adjust the regular expression that works with different contig headers in the ~/.phyluce.conf file.

-b

**Comment by stuartwillis** [comment]
Hi Brant,

We have Trinity 2.2.0 installed on our cluster, though I was going to give Binpacker and Transabyss a shot too. Was there an older version of Trinity you were fonder of than the 2.0.6 that installs with phyluce?

These are the fasta headers for the last assembly it matches and the one following it:
>TRINITY_DN11697_c0_g1_i1 len=1470 path=[1473:0-1469] [-1, 1473, -2]
>TRINITY_DN11697_c1_g1_i1 len=221 path=[397:0-220] [-1, 397, -2]

It's not the first set from the same Trinity supercluster.

I found phyluce.conf in miniconda2/envs/phyluceENV/config (there was no .phyluce.conf in ~/). Indeed the Trinity fasta headers appear to have changed over time (ugh), so the regex may be the problem (not sure why it happens part way though the file though).

This is not a regex format I'm overly familiar with: comp\d+_c\d+_seq\d+|c\d+_g\d+_i\d+|TR\d+\|c\d+_g\d+_i\d+ 

but it seems like this might work (until they change the format...again):
comp\d+_c\d+_seq\d+|c\d+_g\d+_i\d+|TR\d+\|c\d+_g\d+_i\d+|TRINITY_DN\d+_c\d+_g\d+_i\d+

Is it getting applied to extract the contig name? Maybe it would just be easier to sed the fasta headers into something that matches the format from 2.0.6 (e.g. TR1|c0_g1_i1)? Thanks!

**Comment by brantfaircloth** [comment]
I think it's the change to trinity.  See also #71.  It looks like your suggested change of `TRINITY_DN\d+_c\d+_g\d+_i\d+` will capture what's needed to have the program run.

As you guessed, the regex is basically being run to extract the contig name (which then goes into a database to associate a given contig name with a given UCE locus).  The problem is that the Trinity developers change their contig naming scheme a lot (for a variety of reasons).  My solution was to make it reasonably easy (editing `phyluce.conf`) to "fix" that on your own.

In terms of a fix, you should be able to get by with adding:
```
comp\d+_c\d+_seq\d+|c\d+_g\d+_i\d+|TR\d+|c\d+_g\d+_i\d+|TRINITY_DN\d+_c\d+_g\d+_i\d+
```
to `phyluce.conf`.  You can do that either in the `miniconda2/envs/phyluceENV/config` or by copying that file to `~/.phyluce.conf` and editing the correct line.

In terms of other assembly programs, you'll likely run into some of the same problems (harvesting contig name from the assembly output).  And, in addition to Binpacker and Transabyss, you may want to look at https://github.com/juliema/aTRAM, which is about to be released as version 2 (later this week?) and which also looks very promising.

**Comment by stuartwillis** [comment]
Thanks Brant. Copying to .phyluce.conf is simply meant to supercede the conda file, I assume.

So to use other assemblers, should I just add another line in phyluce.conf under [headers], e.g.
binpacker:BINPACKER\d+\.\d+\.\d+
for >BINPACKER.0.1

Or, how is the script determining which regex to apply?

**Comment by brantfaircloth** [comment]
yes, theoretically, that should work.  the script is just generically concatenating all regular expressions for the `[Header]` section with regex OR `|`.

**Comment by stuartwillis** [comment]
gotcha. Thanks!

**Comment by brantfaircloth** [comment]
oh, also meant to say that anything in `~/.phyluce.conf` should override the defaults (you don't even need to copy the entire file - just the sections you want to override).

**Comment by JimStarrett** [comment]
Hi Brant

I seem to be running into a similar problem. When matching contigs to probes I get the error message below that seems related to my headers in my contigs file. The contigs were generated with abyss on a server and the contig files downloaded to my computer where I'm running phyluce. Match_contigs_to_probes has run fine with other data (eg velvet assemblies generated within phyluce).

Thanks!
Jim

Sequence headers:
>3 88 64
TTATATGATAAGGAAGTTATTAAAACCTTTTATTTTTACATATACATATATTTTGAAATTTTTACTTATTTTTTACTCTAAACTTAAA
>4 129 144
GCCATCATTAGGACGAGAGATTTCTTCTTTTTTCTCTCACTCCACACACCCTTCCACACATTTTGAAGTTTATATCACCTATGGTGTCGGTAATCACGCTATTTGTCTCCAGTGCTTTAGCCGTGCATG
>5 142 170
GGAGTTCGGAAAACCATTCATTTTTTTCTTCTTCCTATTCGGTGCGAGTACCATTTCATTTGACTTGCTTGTCATCTCACAAACCGGAAGAAAGATTTCATTTGTTCTGTGTAAATTGTGAAAAAATATTGAATCAAGCTCC


Error message:
  File "/Users/Aliatypus/anaconda2/bin/phyluce_assembly_match_contigs_to_probes", line 345, in <module>
    main()
  File "/Users/Aliatypus/anaconda2/bin/phyluce_assembly_match_contigs_to_probes", line 285, in main
    raise EnvironmentError("lastz: {}".format(lztstderr))
EnvironmentError: lastz: FAILURE: bad fasta character in /Users/Aliatypus/Desktop/UCE_Atypoides/abyss_assemblos_70/contigs/Ahad-MY459.contigs.fasta, >1308584 2226 34650 476259+,1308600+,1304778-: M

**Comment by brantfaircloth** [comment]
hi jim,

This is happening because the abyss contigs generated by the versions of abyss you used (on a server and not through the phyluce code) have not been cleaned of degenerate base calls.  this normally happens automatically when you run the phyluce abyss code, but not when you run abyss on its own, then phyluce.  the other thing to keep in mind is that these degenerate base calls by abyss are often not correct (see https://doi.org/10.1101/255752). the only sort of easy way to fix is to find a program to convert these, write one, or modify what’s in the phyluce code that does this to work somewhat independently of phyluce.

**Comment by JimStarrett** [comment]
I see! Thanks for the reply and link to the paper, looks like a good one!
